mux/tree.go

```diff
@@ -87,6 +87,7 @@ func (t Tree) Match(path string) (Node, middleware.Middleware, context.Params, s
 				if len(orphanMatches) > 0 {
 					for i := 0; i < len(orphanMatches); i++ {
 						m = orphanMatches[i].node.Middleware().Merge(m)
+						m.Reverse()
```

Comment:
By doin `m.Reverse()` inside a loop you revers it every iteration
in the end depending of the `orphanMatches` slice length, odd or even resulting in reversed or not reversed order

also as I said no need to do another loop, if you need to refer which I do not think we need to. just change the logic in the for loop to integrate from end of slice.

Reply:
No matter how I tweaked the for loop I could not get the result achieved with a separate reverse.
Maybe you can provide an example working as described ?

Reply:
Can you please explain what is it that you are trying achieve ? Maybe I dont understand what you want to do.

Reply:
And if we want to fix here anything is the middleware order of nodes from the same tree level that are applied to route node.

This is the order middleware was applied
```go
	router.USE(http.MethodGet, "/", mockFastHTTPMiddleware("m1->"))
	router.USE(http.MethodGet, "/", mockFastHTTPMiddleware("m2->"))
	router.USE(http.MethodGet, "/x", mockFastHTTPMiddleware("mx1->"))
	router.USE(http.MethodGet, "/x", mockFastHTTPMiddleware("mx2->"))
	router.USE(http.MethodGet, "/x/{param}", mockFastHTTPMiddleware("mparam1->"))
	router.USE(http.MethodGet, "/x/{param}", mockFastHTTPMiddleware("mparam2->"))
	router.USE(http.MethodGet, "/x/y", mockFastHTTPMiddleware("mxy1->"))
	router.USE(http.MethodGet, "/x/y", mockFastHTTPMiddleware("mxy2->"))
```
And this would be the fix:
```diff
- m1->m2->mx1->mx2->mxy1->mxy2->mparam1->mparam2->handler
+ m1->m2->mx1->mx2->mparam1->mparam2->mxy1->mxy2->handler
```

Note the change `mxy1->mxy2->mparam1->mparam2` -> `mparam1->mparam2->mxy1->mxy2`. 

But this can not be easily achieved, and i suppose we will not do it.

Reply:
Yes, exactly. This is the issue in most of the tests:

```
--- FAIL: TestNodeApplyMiddlewareDeepStatic (0.00s)
    nethttp_test.go:500: Use middleware error m1m3m2x
--- FAIL: TestFastHTTPNodeApplyMiddlewareDeepWildcardStaticMix (0.00s)
    fasthttp_test.go:483: Use middleware error m1->m2->mx1->mx2->mxy1->mxy2->mparam1->mparam2->handler
--- FAIL: TestNodeApplyMiddleware (0.00s)
    nethttp_test.go:432: Use middleware error m2m1y
    nethttp_test.go:446: Use middleware error m3m1x
--- FAIL: TestFastHTTPNodeApplyMiddlewareDeepStatic (0.00s)
    fasthttp_test.go:458: Use middleware error m1m3m2x
--- FAIL: TestFastHTTPNodeApplyMiddleware (0.00s)
    fasthttp_test.go:402: Use middleware error m2m1y
    fasthttp_test.go:412: Use middleware error m3m1x

```

I think we need to define what we call essential functionality for a router to be meeting basic needs. This example is more of a scientific proof but all the others not passing are basically essential to call it working as expected. In more than 50% of the cases it's a wildcard route and middleware for wildcard and static so if we have to guess the order or document that it will be reversed always that makes it unpredictable even for basic intended usage.